runtime/src/secondary_index.rs

```diff
@@ -160,163 +124,102 @@ impl<SecondaryIndexEntryType: SecondaryIndexEntry + Default + Sync + Send>
-                                key, inner_key, slot);
-                            }
-                        }
-                        w_slot_set.is_empty()
-                    } else {
-                        false
-                    }
-                });
+        let now = solana_sdk::timing::timestamp();
+        let last = self.stats.last_report.load(Ordering::Relaxed);
+        let should_report = now.saturating_sub(last) > 1000
+            && self.stats.last_report.compare_exchange(
+                last,
+                now,
+                Ordering::Relaxed,
+                Ordering::Relaxed,
+            ) == Ok(last);
+
+        if should_report {
+            datapoint_info!(
+                self.metrics_name,
+                ("num_secondary_keys", self.index.len() as i64, i64),
+                (
+                    "num_inner_keys",
+                    self.stats.num_inner_keys.load(Ordering::Relaxed) as i64,
+                    i64
+                ),
+                (
+                    "num_reverse_index_keys",
+                    self.reverse_index.len() as i64,
+                    i64
+                ),
+            );
+        }
+    }
 
-            // Check if `key` is empty
-            if is_inner_key_empty {
-                // Write lock on `inner_key_entry` above through the `entry`
-                // means nobody else has access to this lock at this time,
-                // so this check for empty -> remove() is atomic
-                inner_key_map.remove_key_if_empty(inner_key);
-                inner_key_map.is_empty()
-            } else {
-                false
-            }
-        } else {
-            false
+    // Only safe to call from `remove_by_inner_key()` due to asserts
+    fn remove_index_entries(&self, outer_key: &Pubkey, removed_inner_key: &Pubkey) {
+        let is_outer_key_empty = {
+            let inner_key_map = self
+                .index
+                .get_mut(&outer_key)
+                .expect("If we're removing a key, then it must have an entry in the map");
+            // If we deleted a pubkey from the reverse_index, then the corresponding entry
+            // better exist in this index as well or the two indexes are out of sync!
+            assert!(inner_key_map.value().remove_inner_key(&removed_inner_key));
+            inner_key_map.is_empty()
         };
 
         // Delete the `key` if the set of inner keys is empty
-        if is_key_empty {
+        if is_outer_key_empty {
             // Other threads may have interleaved writes to this `key`,
             // so double-check again for its emptiness
-            if let Occupied(key_entry) = self.index.entry(*key) {
+            if let Occupied(key_entry) = self.index.entry(*outer_key) {
                 if key_entry.get().is_empty() {
                     key_entry.remove();
                 }
             }
```

Comment:
ah shoot, also I don't think this check for empty -> deletion triggered from  is safe because we're no longer accounting for slots.

For instance, if:

1) clean calls `handle_dead_keys(pubkey)` where the mint for `pubkey` is `M`,
2) a new block `B` inserts `pubkey` with mint also `M` into `AccountsIndex`, then calls `update_secondary_indexes()`: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_index.rs#L1153, the new insert here only checks `M` exists in the secondary index set (doesn't add a unique entry for block `B` if `M` already exists)
3) Clean now finishes removing `pubkey` from the secondary mint index `M` and comes into this check, and removes this key entirely, even though block `B` is keeping this mapping alive.

I think if we pass the atomic `refcount` from the AccountsIndex here, it will solve this issue.